File: composer-show-template/src/main/java/io/piano/android/composer/showtemplate/ShowTemplateController.java
package io.piano.android.composer.showtemplate;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.View;
import android.webkit.WebView;

import androidx.fragment.app.FragmentActivity;

import io.piano.android.composer.model.DelayBy;
import io.piano.android.composer.model.Event;
import io.piano.android.composer.model.events.ShowTemplate;

import static android.view.View.VISIBLE;

public class ShowTemplateController {

    private WebView webView;
    private ShowTemplateDialogFragment fragment;

    private ShowTemplateController() {
    }

    private ShowTemplateController(WebView webView) {
        this.webView = webView;
    }

    private ShowTemplateController(ShowTemplateDialogFragment fragment) {
        this.fragment = fragment;
    }

    public void reloadWithToken(String userToken) {
        String jsSnippet = "javascript:piano.reloadTemplateWithUserToken('" + userToken + "')";

        if (fragment != null) {
            fragment.getWebView().loadUrl(jsSnippet);
        } else {
            webView.loadUrl(jsSnippet);
        }
    }

    public static ShowTemplateController show(final Context context, final Event<ShowTemplate> showTemplateEvent) {
        return show(context, showTemplateEvent, null);
    }

    public static ShowTemplateController show(final Context context, final Event<ShowTemplate> showTemplateEvent, final Object javascriptInterface) {
        switch (showTemplateEvent.eventData.displayMode) {
            case MODAL:
                return showModal(context, showTemplateEvent, javascriptInterface);
            case INLINE:
                return showInline(context, showTemplateEvent, javascriptInterface);
        }

        return null;
    }

    private static ShowTemplateController showModal(final Context context, Event<ShowTemplate> showTemplateEvent, Object javascriptInterface) {
        if (!(context instanceof FragmentActivity)) {
            return null;
        }

        final ShowTemplateDialogFragment fragment = new ShowTemplateDialogFragment();
        ShowTemplate showTemplate = showTemplateEvent.eventData;

        boolean cancelable = showTemplate.showCloseButton;
        if (!cancelable) {
            fragment.setCancelable(false);
        }

        if (javascriptInterface != null) {
            fragment.setJavascriptInterface(javascriptInterface);
        }

        Bundle args = new Bundle();
        args.putString("url", showTemplate.url);
        args.putString("trackingId", showTemplateEvent.eventExecutionContext.trackingId);
        fragment.setArguments(args);

        boolean showImmediately = true;
        if (showTemplate.delayBy.type == DelayBy.DelayType.TIME) {
            final int delaySecs = showTemplate.delayBy.value;
            if (delaySecs > 0) {
                showImmediately = false;
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        FragmentActivity activity = (FragmentActivity) context;
                        if (!activity.isFinishing()) {
                            fragment.show(activity.getSupportFragmentManager(), "ShowTemplateDialogFragment");
                        }
                    }
                }, delaySecs * 1000);
            }
        }

        if (showImmediately) {
            fragment.show(((FragmentActivity) context).getSupportFragmentManager(), "ShowTemplateDialogFragment");
        }

        return new ShowTemplateController(fragment);
    }

    private static ShowTemplateController showInline(final Context context, final Event<ShowTemplate> showTemplateEvent, Object javascriptInterface) {
        ShowTemplate showTemplate = showTemplateEvent.eventData;
        if (TextUtils.isEmpty(showTemplate.containerSelector)) {
            return null;
        }

        int id = context.getResources().getIdentifier(showTemplate.containerSelector, "id", context.getPackageName());
        if (id == 0) {
            return null;
        }

        View view = ((Activity) context).getWindow().getDecorView().findViewById(id);
        if (!(view instanceof WebView)) {
            return null;
        }

        final WebView webView = (WebView) view;
        InitWebViewHelper.initWebView(null, webView, javascriptInterface, showTemplateEvent.eventExecutionContext.trackingId);

        boolean showImmediately = true;
        if (showTemplate.delayBy.type == DelayBy.DelayType.TIME) {
            final int delaySecs = showTemplate.delayBy.value;
            if (delaySecs > 0) {
                showImmediately = false;
                new Handler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        if (!((Activity) context).isFinishing()) {
                            webView.loadUrl(showTemplate.url);
                            webView.setVisibility(VISIBLE);
                        }
                    }
                }, delaySecs * 1000);
            }
        }

        if (showImmediately) {
            webView.loadUrl(showTemplate.url);
            webView.setVisibility(VISIBLE);
        }

        return new ShowTemplateController(webView);
    }

}

File: composer-show-template/src/main/java/io/piano/android/composer/showtemplate/ShowTemplateController.kt
package io.piano.android.composer.showtemplate

import android.annotation.SuppressLint
import android.os.Handler
import android.os.Looper
import android.webkit.WebView
import androidx.annotation.UiThread
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import io.piano.android.common.DeviceIdProvider
import io.piano.android.composer.model.Event
import io.piano.android.composer.model.events.ShowTemplate
import io.piano.android.showhelper.BaseShowController

/**
 * A controller for displaying a template using the Piano Composer API.
 *
 * This controller is responsible for showing a template specified by the [ShowTemplate] event.
 * It extends the [BaseShowController] with [ShowTemplate] data and [ComposerJs] JavaScript
 * interface. It handles the template display, delay, and interaction with the JavaScript interface.
 *
 * @param event The [Event] containing the [ShowTemplate] data to be displayed.
 * @param deviceIdProvider The [DeviceIdProvider] instance
 * @param jsInterface The optional [ComposerJs] JavaScript interface for communication with the
 *                    displayed template. If not provided, a default instance of [ComposerJs] will
 *                    be used.
 * @param jsInterfaceName The optional name under which the JavaScript interface is exposed to the
 *                        template. If not provided, defaults to [JAVASCRIPT_INTERFACE].
 */
public class ShowTemplateController(
    event: Event<ShowTemplate>,
    deviceIdProvider: DeviceIdProvider,
    jsInterface: ComposerJs? = null,
    private val jsInterfaceName: String = JAVASCRIPT_INTERFACE,
) : BaseShowController<ShowTemplate, ComposerJs>(event.eventData, jsInterface ?: ComposerJs()) {
    // Private properties
    private val trackingId = event.eventExecutionContext.trackingId

    // Overrides from BaseShowController
    override val url: String = event.eventData.url ?: "about:blank"
    override val additionalHttpHeaders: Map<String, String> =
        mapOf(DeviceIdProvider.DEVICE_ID_HEADER to deviceIdProvider.deviceId)
    override val fragmentTag: String = FRAGMENT_TAG
    override val fragmentProvider: () -> ShowTemplateDialogFragment = {
        ShowTemplateDialogFragment(url, trackingId, additionalHttpHeaders)
    }

    override fun WebView.configure(): Unit = prepare(null, jsInterface, trackingId, jsInterfaceName)

    override fun processDelay(activity: FragmentActivity, showFunction: () -> Unit) {
        val func: () -> Unit = {
            if (!activity.isFinishing) {
                showFunction()
            }
        }
        eventData.delayBy.apply {
            if (isDelayedByTime) {
                val handler = Handler(Looper.getMainLooper())
                activity.lifecycle.addObserver(
                    object : DefaultLifecycleObserver {
                        override fun onPause(owner: LifecycleOwner) {
                            handler.removeCallbacksAndMessages(null)
                        }
                    },
                )
                handler.postDelayed(func, value * 1000L)
            } else {
                func()
            }
        }
    }

    /**
     * Closes the template via the JavaScript interface `close` function.
     *
     * @param data Optional data to be passed to the `close` function of the JavaScript interface.
     */
    @Suppress("unused") // Public API.
    @UiThread
    override fun close(data: String?): Unit = jsInterface.close(data)

    /**
     * Sends a reload command to the template. This should be called when the user token has changed.
     *
     * @param userToken The updated user token to be sent to the JavaScript interface.
     */
    @Suppress("unused") // Public API.
    public fun reloadWithToken(userToken: String): Unit = jsInterface.updateToken(userToken)

    internal companion object {
        private const val FRAGMENT_TAG = "ShowTemplateDialogFragment"
        internal const val JAVASCRIPT_INTERFACE = "PianoAndroid"

        // Internal function for WebView preparation
        @SuppressLint("SetJavaScriptEnabled")
        internal fun WebView.prepare(
            dialogFragment: ShowTemplateDialogFragment? = null,
            javascriptInterface: ComposerJs?,
            trackingId: String,
            jsInterfaceName: String = JAVASCRIPT_INTERFACE,
        ) {
            settings.javaScriptEnabled = true
            val jsInterface = javascriptInterface ?: ComposerJs()
            jsInterface.init(dialogFragment, this, trackingId)
            addJavascriptInterface(jsInterface, jsInterfaceName)
        }
    }
}
